src/main/java/com/limechain/trie/structure/TrieStructure.java

```diff
@@ -494,18 +500,42 @@ public <U> boolean structurallyEquals(TrieStructure<U> other) {
         }
     }
 
-    public boolean removeValueAtKey(Nibbles key) {
+    public boolean deleteNodeAt(Nibbles key) {
         return switch (existingNodeInner(key)) {
             case ExistingNodeInnerResult.NotFound ignored -> false;
             case ExistingNodeInnerResult.Found found -> clearNodeValue(found.nodeIndex);
         };
     }
 
-    public boolean clearNodeValue(TrieNodeIndex nodeIndex) {
-        return clearNodeValue(nodeIndex.getValue());
+    public void deleteNodesRecursively(TrieNodeIndex nodeIndex, Long limit, AtomicInteger deleted) {
+        deleteNodesRecursively(nodeIndex.getValue(), limit, deleted);
     }
 
-    boolean clearNodeValue(int nodeIndex) {
+    private void deleteNodesRecursively(int nodeIndex, Long limit, AtomicInteger deleted) {
+        TrieNode<T> trieNode = getNodeAtIndexInner(nodeIndex);
+
+        TrieNode.Parent parent = trieNode.parent;
+        if (parent == null) {
+            nodes.remove(nodeIndex);
+             deleted.incrementAndGet();
+             return;
+        }
+        TrieNode<T> parentNode = getNodeAtIndexInner(parent.parentNodeIndex());
+
+        for (Integer childrenIndex : trieNode.childrenIndices) {
+            if (limit != null && deleted.get() >= limit) {
+                return;
+            }
+            if (childrenIndex != null) {
+                deleteNodesRecursively(childrenIndex, limit, deleted);
+            }
+        }
+
+        parentNode.childrenIndices[parent.childIndexWithinParent().asInt()] = null;
+        nodes.remove(nodeIndex);
+    }
+
+    private boolean clearNodeValue(int nodeIndex) {
```

Comment:
I'd rather call this `deleteNode` or `removeNode` since that's the actual brute force removing of the node with the index.